src/estimagic/inference/bootstrap_samples.py

```diff
@@ -19,17 +26,45 @@ def get_bootstrap_indices(data, rng, cluster_by=None, n_draws=1000):
 
     """
     n_obs = len(data)
-    if cluster_by is None:
-        bootstrap_indices = list(rng.integers(0, n_obs, size=(n_draws, n_obs)))
+
+    if weights is None:
+
+        if cluster_by is None:
+            bootstrap_indices = list(rng.integers(0, n_obs, size=(n_draws, n_obs)))
+        else:
+            clusters = data[cluster_by].unique()
+            drawn_clusters = rng.choice(
+                clusters, size=(n_draws, len(clusters)), replace=True
+            )
+
+            bootstrap_indices = _convert_cluster_ids_to_indices(
+                data[cluster_by], drawn_clusters
+            )
+
     else:
-        clusters = data[cluster_by].unique()
-        drawn_clusters = rng.choice(
-            clusters, size=(n_draws, len(clusters)), replace=True
-        )
 
-        bootstrap_indices = _convert_cluster_ids_to_indices(
-            data[cluster_by], drawn_clusters
-        )
+        if cluster_by is None:
+            bootstrap_indices = list(
+                rng.choice(
+                    n_obs,
+                    size=(n_draws, n_obs),
+                    replace=True,
+                    p=data[weights] / data[weights].sum(),
+                )
+            )
```

Comment:
It would be more readable if you compute the vector of probabilities outside of these if conditions. If `weights` is `None`, the vector of probabilities could be `None`. In this case, you could also write the same code for both cases:

```python
probs = _calculate_bootstrap_indices_probs(data, weight_by, cluster_by) 

bootstrap_indices = list(
    rng.choice(
        n_obs,
        size=(n_draws, n_obs),
        replace=True,
        p=probs,
    )
)
```

This will be slower than `np.integers(n_obs, ...)` but I think this is a trade-off we would want to make here.